Sources/SourceControl/GitRepository.swift

```diff
@@ -349,6 +349,21 @@ public class GitRepository: Repository, WorkingCheckout {
         try runCommandQuietly([Git.tool, "-C", path.asString, "reset", "--hard", revision.identifier])
     }
 
+    /// Returns true if a revision exists.
+    public func exists(revision: Revision) -> Bool {
+        do {
+           _ = try runCommandQuietly([Git.tool, "-C", path.asString, "rev-parse", "--verify", revision.identifier])
+        } catch {
+            return false
+        }
+        return true
+    }
+
+    public func checkout(newBranch: String) throws {
+        precondition(isWorkingRepo, "This operation should run in a working repo.")
+        try runCommandQuietly([Git.tool, "-C", path.asString, "checkout", "-b", newBranch])
```

Comment:
Do we also need to guard against name conflicts here?  For example, creating a new branch with the same name as an existing tag creates the branch without warning.  But then whenever that branch is mentioned again, git emits a warning about being it ambiguous (and then the branch is favored).